brokers/artifacts/broker.go

```diff
@@ -0,0 +1,179 @@
+//
+// Copyright (c) 2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package artifacts
+
+import (
+	"fmt"
+	"path/filepath"
+
+	jsonrpc "github.com/eclipse/che-go-jsonrpc"
+	"github.com/eclipse/che-plugin-broker/common"
+	"github.com/eclipse/che-plugin-broker/model"
+	"github.com/eclipse/che-plugin-broker/utils"
+)
+
+const errorNoExtFieldsTemplate = "Field 'extensions' is not found in the description of the plugin '%s'"
+
+// Broker is used to process Che plugins
+type Broker struct {
+	common.Broker
+	ioUtils utils.IoUtil
+	rand    common.Random
+}
+
+// NewBroker creates Che broker instance
+func NewBroker(localhostSidecar bool) *Broker {
+	return &Broker{
+		Broker:  common.NewBroker(),
+		ioUtils: utils.New(),
+		rand:    common.NewRand(),
+	}
+}
+
+func (b *Broker) fail(err error) error {
+	b.PubFailed(err.Error())
+	b.PubLog(err.Error())
+	return err
+}
+
+// PushEvents sets given tunnel as consumer of broker events.
+func (b *Broker) PushEvents(tun *jsonrpc.Tunnel) {
+	b.Broker.PushEvents(tun, model.BrokerStatusEventType, model.BrokerResultEventType, model.BrokerLogEventType)
+}
+
+// Start downloads metas from plugin registry for specified
+// pluginFQNs and then executes plugins metas processing and sending data to Che master
+func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) error {
+	b.cleanupPluginsDirectory(b.ioUtils)
+
+	pluginMetas, err := utils.GetPluginMetas(pluginFQNs, defaultRegistry, b.ioUtils)
+	if err != nil {
+		return b.fail(fmt.Errorf("Failed to download plugin meta: %s", err))
+	}
+	defer b.CloseConsumers()
+	b.PubStarted()
```

Comment:
Should not be PubStarted a first thing broker do after start?
It's just a question since it's not clear how it works and it's not documented.
If it worked in the same way before (I guess it did) it's OK to leave as is.

Reply:
Good point, that makes sense to me. AFAICT `PubStarted` *should* happen immediately, I'll make the change and re-test. (I don't know why it was done the other way in the unified broker)